### PR TITLE
Refactored _submitJobs

### PR DESCRIPTION
**Pull request by maikenp**
Sorry, made a squash mesh. New PR with all proposed changes included. 

Have tested this for my usecase.

### CODE REVIEW

**Comment by maikenp** [comment]
My implementation uses submitJobs like this - this is an extract of the arc-galaxy job runner. 

The ```self.prepare_job(job_wrapper, self.arcjob)``` populates the ARCJob object so that it contains the jobdescription string in addition to queue, inputFiles and outputFiles. It will contain all that is needed for proper processing, eventually also runtimes (does not yet). 

By setting ```parseDescs=False, processDescs=False``` in the call to submitJobs I skip the parsing and unparsing using arclib as this is now isolated in separate methods. See the PR. 

```
def queue_job(self, job_wrapper):
        """ When a tool is submitted for execution in galaxy """

        """ Create job script and submit it to arc - set arc cluster variable """
        job_destination = job_wrapper.job_destination
        galaxy_job_id = job_wrapper.job_id

        """ Set the cluster to submit the job to - extracted from the job_destination parameters in job_conf.xml """
        self.arc.set_job_cluster(job_destination.params["arc_url"])
        
        """ Prepare and submit job to arc """
        self.prepare_job(job_wrapper, self.arcjob)
        arcjobs = {0:self.arcjob}

        token = job_wrapper.get_job().user.get_oidc_tokens(self.provider_backend)['access']
        arc_url = job_wrapper.job_destination.params['arc_url']
        self.arcrest = ArcRest.getClient(url=arc_url, version="1.1", proxypath=PROXYPATH, token=token, impls={"1.1":ArcRest})
        
        delegationID = self.arcrest.createDelegation(token=token)
        queue = self.arcjob.Queue
        resultList = self.arcrest.submitJobs(queue, arcjobs=arcjobs, token=token, delegationID=delegationID, parseDescs=False, processDescs=False, matchDescs=False)
        arc_job_id = ''
        for arcid, state in resultList:
            arc_job_id = arcid

        if not arc_job_id:
            log.error("Job creation failure.  No Response from ARC")
            job_wrapper.fail("Not submitted")
        else:
            self.arc.job_mapping[galaxy_job_id]=arc_job_id
            # Create an object of AsynchronousJobState and add it to the monitor queue.
            ajs = AsynchronousJobState(files_dir=job_wrapper.working_directory, job_wrapper=job_wrapper, job_id=galaxy_job_id, job_destination=job_destination)
            self.monitor_queue.put(ajs)

```

File: src/pyarcrest/arc.py [hunk 1]
```diff
@@ -242,13 +243,15 @@ def newDelegation(self, token=None, proxypath=None):
...
+            headers["Authorization"] = f"Bearer {token}"
             params["type"] = "jwt"
 
+        endpoint = f"{self.apiPath}/delegations"
         resp = self._request("POST", "/delegations", headers=headers, params=params, token=token)
         respstr = resp.read().decode()
 
         if token:
-            if resp.status != 200:
+            if resp.status != 201:
```
Comment (jakobmerljak) [inline src/pyarcrest/arc.py:254]: If 201 is returned for tokens, we can again report the fix for the [this](https://www.nordugrid.org/arc/arc7/tech/rest/rest.html#jwt-delegation) section of documentation. You can also submit a separate pull request for pyarcrest to fix this.

**Comment by jakobmerljak** [comment]
Contrary to my initial suggestion, I don't think that using *submitJobs* is the best way of doing this for your use case. There are several reasons for that, especially from the perspective of the interface of *submitJobs*.

The process of job submission to ARC is quite involved and requires several operations on the REST API. The main purpose of *submitJobs* is to combine those in a single method, suitable for general use cases (e. g. it fits well with aCT and ARC client).

For completeness sake I will list the possible steps that are performed in the general process of job submission:
- creating new delegation for job
- job description parsing
- job description processing (user xRSL to server xRSL)
- matchmaking some job description parameters
- adding delegation and queue to job description in for API 1.0
- extracting local input file information from description and uploading the files to ARC CE
- unparsing processed/modified job description to ADL/xRSL
- submitting job descriptions to ARC CE

Depending on the integration, some steps can be optional as they might be done elsewhere using other parts of the library (e. g. creation of delegation or matchmaking in case of more sophisticated job brokering mechanisms).

Your use case is quite different as it generates job descriptions from several parameters instead of parsing them. It therefore doesn't make sense for *submitJobs* to parse and process the generated descriptions and regenerate them again. This is reflected in your proposed change which uses *ARCJob* objects and additional flag as an alternative set of parameters.

There are several reasons why I think this changed API would be suboptimal. Firstly, my intent is not to use *ARCJob* objects in the *ARCRest* class. I started working on *ARCClient* class which would have the same interface except that *ARCJob* objects would be used as parameters and return values, but I haven't had time to continue working on that yet.

More importantly, I think that the new interface of *submitJobs* would not be as intuitive anymore. The problem is that a single method would now submit jobs from two quite different "sources" which have to be handled, prepared and perceived differently. One is a list of description strings and another is a list of *ARCJob* objects. I think this makes the method more complicated and harder to understand (it is quite loaded already as it is).

The method could be used in two different ways. The first is to just pass in the job descriptions. The other one is to prepare *ARCJob* objects in an arbitrary way and pass them in. There is even a third option where job descriptions can be passed in *ARCJob* objects with the `parseDescs=True` parameter.

Generally from the API perspective, I think that two distinct methods instead of a single one would have to be used for the two different use cases that we are dealing with here. So, a second method could be added that submits *ARCJob* objects. But, I also think that there are better options than adding another method to the pyarcrest API. I think that the integrated procedure of job submission could be better implemented in the Galaxy integration instead of pyarcrest. At least for now, as this is a single such use case. I would like to keep the API as small and as simple as possible, especially since pyarcrest is still not finished yet. Later, if there are more such use cases, we could consider implementing them as well.

You could implement submission somewhat like this:
``` python
def queue_job(self, job_wrapper):
    """ When a tool is submitted for execution in galaxy """

    """ Create job script and submit it to arc - set arc cluster variable """
    job_destination = job_wrapper.job_destination
    galaxy_job_id = job_wrapper.job_id

    """ Set the cluster to submit the job to - extracted from the job_destination parameters in job_conf.xml """
    self.arc.set_job_cluster(job_destination.params["arc_url"])

    """ Prepare and submit job to arc """
    self.prepare_job(job_wrapper, self.arcjob)
    arcjobs = {0:self.arcjob}

    token = job_wrapper.get_job().user.get_oidc_tokens(self.provider_backend)['access']
    arc_url = job_wrapper.job_destination.params['arc_url']

    # proxypath is ignored if you are using token
    self.arcrest = ArcRest.getClient(url=arc_url, version="1.1", token=token, impls={"1.1":ArcRest})

    # token parameter isn't necessary, unless there is a bug
    delegationID = self.arcrest.createDelegation()

    queue = self.arcjob.Queue

    bulkdesc = "<ActivityDescriptions>"
    for job in arcjobs:
        bulkdesc += job.descstr
    bulkdesc += "</ActivityDescriptions>"

    results = self.arcrest.createJobs(bulkdesc, queue, delegationID)
    
    if isinstance(results[0], ARCHTTPError):  # submission error
        log.error("Job creation failure.  No Response from ARC")
        job_wrapper.fail("Not submitted")
    else:  # successful submission
        jobid, _ = result[0]
        # beware! this means 1 worker, no timeout and default upload buffer
        errors = self.arcrest.uploadJobFiles(jobid, self.arcjob.inputFiles)
        if errors[0]:  # input upload error
            log.error("Job creation failure. No Response from ARC")
            job_wrapper.fail("Not submitted")
        else:  # successful input upload
            self.arc.job_mapping[galaxy_job_id]=arc_job_id
            # Create an object of AsynchronousJobState and add it to the monitor queue.
            ajs = AsynchronousJobState(files_dir=job_wrapper.working_directory, job_wrapper=job_wrapper, job_id=galaxy_job_id, job_destination=job_destination)
            self.monitor_queue.put(ajs)
```

Code looks similar to parts of the current *submitJobs*. I think it is overall cleaner to handle some of this explicitly, both on API consumer and provider sides. Otherwise, the interface and implementation becomes more complex. But, when I get back to experimenting with a higher level API with *ARCJob* objects, it might become simpler for your use case.

**Comment by maikenp** [comment]
Thanks for your comments, and I agree, I was not convinced that submitJobs now either handle descs lists or arcjob objects, seems messy to handle. And yes, I get the point that submitJobs was supposed to entail all the needed steps. 

The main goal of my refactoring was in the end to ensure that I did not have to use arclib to parse/unparse the job description. And as you suggest, by just adding the extra bulkdesc part I can use the createJobs directly. That is just perfect. 

So yes, this is much simpler. And I am in this way free to use ARCJob the way I wish without enforcing that in pyarcrest. 

For ARCJob - could we add a runtimes list too (or was it dict?) then we have I think all we need in the ARCJob object itself from the job description.

**Comment by jakobmerljak** [comment]
> Thanks for your comments, and I agree, I was not convinced that submitJobs now either handle descs lists or arcjob objects, seems messy to handle. And yes, I get the point that submitJobs was supposed to entail all the needed steps.
> 
> The main goal of my refactoring was in the end to ensure that I did not have to use arclib to parse/unparse the job description. And as you suggest, by just adding the extra bulkdesc part I can use the createJobs directly. That is just perfect.

I gave you a wrong suggestion to override and use *submitJobs* for your use case initially. I am glad we figured it out.

> For ARCJob - could we add a runtimes list too (or was it dict?) then we have I think all we need in the ARCJob object itself from the job description.

Runtimes are given as a list of strings in arclib and they are processed as such in pyarcrest currently. Yes, I think we can add them to ARCJob as a list attribute "runtimes".

If you have commits handy you are welcome to submit PR.

**Comment by maikenp** [comment]
Decided against this.
